src/qdwh.cc

```diff
@@ -0,0 +1,410 @@
+    using blas::real;
+
+    int64_t max_panel_threads  = std::max(omp_get_max_threads()/2, 1);
+
+    max_panel_threads = get_option<int64_t>( opts, Option::MaxPanelThreads, max_panel_threads );
+
+    int itconv, it;
+
+    real_t L2, sqd, dd, a1, a, b, c;
+    real_t Li, Liconv;
+    real_t conv = 100.;
+    real_t rzero = 0.0, rone = 1.0;
+
+    scalar_t zero = 0.0, one = 1.0;
+    scalar_t alpha, beta;
+
+    real_t normR;
+
+    real_t eps  = std::numeric_limits<real_t>::epsilon();
+    real_t tol1 = 5. * eps;
+    real_t tol3 = pow(tol1, rone/real_t(3.));
+
+    int64_t mt = A.mt();
+    int64_t nt = A.nt();
+
+    int64_t m  = A.m();
+    int64_t n  = A.n();
+
+    if (m < n) {
+        return;
+    }
+
+    int64_t nb   = A.tileMb(0);
+
+    // The QR-based iterations requires QR[A; Id],
+    // W = [A; Id], where size of A is mxn, size of Id is nxn
+    // To avoid having a clean up tile in the middle of the W matrix,
+    // we round up the number of A rows (m),
+    // so that size(W) = m_roundup + n
+    int64_t m_roundup =  ((m + nb - 1) / nb ) * nb;
+    int64_t m_W   = m_roundup + n;
+    int64_t mt_W = mt + nt;
+
+    int nprow, npcol;
+    int myrow, mycol;
+    GridOrder order;
+    A.gridinfo(&order, &nprow, &npcol, &myrow, &mycol);
+
+    // allocate m_W*n work space required for the qr-based iterations QR([A;Id])
+    slate::Matrix<scalar_t> W(m_W, n, nb, nprow, npcol, A.mpiComm());
+    slate::Matrix<scalar_t> Q(m_W, n, nb, nprow, npcol, A.mpiComm());
+    slate::TriangularFactors<scalar_t> T;
+
+    slate::Matrix<scalar_t> Acpy;
+    if (m == n) {
+        // if A is square, we can use H to save a copy of A
+        // this is will not work if H is stored as Hermitian matrix
+        Acpy = H;
+    }
+    else {
+        Acpy = slate::Matrix<scalar_t>(m, n, nb, nprow, npcol, A.mpiComm());
+        Acpy.insertLocalTiles(target);
+    }
+
+    W.insertLocalTiles(target);
+    Q.insertLocalTiles(target);
+
+    // To compute QR[A; Id], allocate W and Q of size m_Wxn
+    // First mt tiles of W contains A matrix in [0:m-1, 0:n-1]
+    // W1 have more rows than A, since we round up the number of rows of W1
+    auto W1 = W.sub(0, mt-1, 0, nt-1);
+    auto Q1 = Q.sub(0, mt-1, 0, nt-1);
+    // W10 have A matrix
+    auto W10 = W1.slice(0, m-1, 0, n-1);
+    auto Q10 = Q1.slice(0, m-1, 0, n-1);
+    // W11 is the last tile which has more rows in case we round up m
+    auto W11 = W.sub(mt-1, mt-1, 0, nt-1);
+
+    // Second nt tiles of W, contain the identity matrix
+    auto W2 = W.sub(mt, mt_W-1, 0, nt-1);
+    auto Q2 = Q.sub(mt, mt_W-1, 0, nt-1);
+
+    // backup A in Acpy to compute H
+    slate::copy(A, Acpy, opts);
+
+    // two norm estimation (largest singular value of A)
+    real_t norm_est = norm2est( A, opts);
+    alpha = 1.0;
+
+    // scale the original matrix A to form A0 which is the initial matrix
+    // for the first iteration
+    add(alpha/scalar_t(norm_est), Acpy, zero, A, opts);
```

Comment:
Unless norm2est modified `A`, doesn't `A` already have the data? Wouldn't it be better to use `scale( alpha/norm_est, A )` instead of `add`?

Reply:
Right, I will use scale.